### PR TITLE
gh-128234: support emscripten and wasi in async contextlib tests, by removing asyncio from contextlib async tests

### PR DESCRIPTION
**Pull request by graingert**
* Issue: gh-128234

### CODE REVIEW

File: Lib/test/test_contextlib_async.py [hunk 1]
```diff
@@ -8,14 +8,32 @@
 
 from test.test_contextlib import TestBaseExitStack
 
-support.requires_working_socket(module=True)
 
-def tearDownModule():
-    asyncio._set_event_loop_policy(None)
+def _run_async_fn(async_fn, /, *args, **kwargs):
+    coro = async_fn(*args, **kwargs)
+    gen = type(coro).__await__(coro)
```
Comment (kumaraditya303) [inline Lib/test/test_contextlib_async.py]: You can directly call `coro.send(None)` here

File: Lib/test/test_contextlib_async.py [hunk 2]
```diff
@@ -8,14 +8,32 @@
...
+    try:
+        gen.send(None)
+    except StopIteration as e:
+        return e.value
+    else:
+        raise AssertionError("coroutine did not stop")
```
Comment (kumaraditya303) [inline Lib/test/test_contextlib_async.py]:
```suggestion
        raise AssertionError("coroutine did not complete")
```

**Review by asvetlov — APPROVED** [review]
Looks good, thanks!

File: Lib/test/test_contextlib_async.py [hunk 3]
```diff
@@ -8,14 +8,32 @@
...
+    finally:
+        gen.close()
 
 
-class TestAbstractAsyncContextManager(unittest.IsolatedAsyncioTestCase):
+def _async_test(async_fn):
+    """Decorator to turn an async function into a test case."""
```
Comment (kumaraditya303) [inline Lib/test/test_contextlib_async.py]: This comment seems wrong, it should be actually `runs async func synchronously`

Comment (graingert) [reply Lib/test/test_contextlib_async.py]:
```suggestion
    """Decorator to turn an async function into a synchronous function"""
```

**Review by kumaraditya303 — APPROVED** [review]

File: Lib/test/test_contextlib_async.py [hunk 4]
```diff
@@ -492,45 +531,27 @@ async def agenfunc():
...
-            return self.run_coroutine(self.__aexit__(*exc_details))
+            return _run_async_fn(self.__aexit__, *exc_details)
 
     exit_stack = SyncAsyncExitStack
     callback_error_internal_frames = [
-        ('__exit__', 'return self.run_coroutine(self.__aexit__(*exc_details))'),
-        ('run_coroutine', 'raise exc'),
-        ('run_coroutine', 'raise exc'),
+        ('__exit__', 'return _run_async_fn(self.__aexit__, *exc_details)'),
+        ('_run_async_fn', 'gen.send(None)'),
```
Comment (kumaraditya303) [inline Lib/test/test_contextlib_async.py]:
```suggestion
        ('_run_async_fn', 'coro.send(None)'),
```

**Pull request merged**